msprime/ancestry.py

```diff
@@ -1146,6 +1146,128 @@ def sim_ancestry(
     )
 
 
+def continue_simulation(
+    ts,
+    time,
+    sample_size=None,
+    continue_nodes=None,
+    **kwargs,
+):
+    """
+    Returns a full tree sequence, by continuing the simulation using msprime
+    after an earlier simulation has been performed. The msprime simulation and slim
+    simulation are joined using :meth:`tskit.tables.TableCollection.union()`.
+
+    This method executes code from the pyslim tutorial  'Following up with more
+    coalescent simulation'
+
+    ``sample_size`` defaults to the number of samples present in the input tree sequence.
+
+    In general, all parameters  excepting ``samples`` are whatever the defaults
+    of ``msprime.simulate`` are; this includes recombination rate, so that if neither
+    ``recombination_rate`` or a ``recombination_map`` are provided, there will be
+    *no* recombination.
+
+    Note that ``Ne`` defaults to ``1.0``; you may want to set it explicitly
+    using kwargs.
+
+    :type ts: :class`tskit.TreeSequence`
+    :param ts: tree sequence used to continue siumlation, moving forward in time
+    :param float time: The desired period of time to continue simulation in msprime,
+        in units of generations
+    :param int sample_size: Sample size for :meth:`msprime.simulate`,
+        in units of haploid samples,
+    :param dict kwargs: Any other arguments to :meth:`msprime.simulate`.
+    :type nodes: :class`numpy.ndarray`
+    :param continue_nodes: node IDs in ts to match up to the lineages in the continued
+        simulation, defaults to the samples
+    :return: The :class:`tskit.TreeSequence` object resulting from the
+        union of the input tree equence and the continued simulatoin
+    :rtype: :class:`tskit.TreeSequence`
+    """
+    if continue_nodes is None:
+        continue_nodes = ts.samples()
+    if sample_size is None:
+        sample_size = len(continue_nodes)
+
+    new_ts = simulate(sample_size=sample_size, end_time=time, **kwargs)
+
+    new_nodes = np.where(new_ts.tables.nodes.time == time)[0]
+
+    if len(new_nodes) > len(continue_nodes):
+        raise RuntimeError(
+            f"""More lineages present in new simulation samples than in the old.. \
+            Decrease sample_size to try again. \
+            Setting sample_size={len(continue_nodes)} \
+            or leaving it to default will ensure compliance."""
+        )
+
+    if len(new_nodes) == 0:
+        oldest_node_time = max(
+            [
+                new_ts.node(new_ts.first().roots[i]).time
+                for i in range(0, len(new_ts.first().roots), 1)
+            ]
+        )
+        raise RuntimeError(
+            f"""No surviving lineages in new simulation at merge time. \
+            Depth of new simulation is {oldest_node_time} - alter parameters \
+            to extend simulation or reduce the simulation time."""
```

Comment:
Oh, right! Forgot about this. So, if there's no nodes with time equal to `time`, that doesn't mean there aren't lineages, it just means that everything's coalesced. There is one lineage still at each position (there is always at least one, since the samples had to inherit from *someone*), but its the root of the tree. What we're doing currently - leaving these out - only gives us the variants that are segregating within the sample at the end. We'd be missing the scribbled edge in this picture:
![trees](https://user-images.githubusercontent.com/1046249/103155287-2e534980-4753-11eb-95fe-640fb5d59d63.jpg)


My first thought is to just add the roots of any trees that have coalesced to `new_nodes`, like so:
```
roots = []
for t in new_ts.trees():
    if t.num_roots == 1:
        roots.append(t.root)
new_nodes.extend(roots)
```
But let's see: then the two nodes we're telling union are the same would have different times, and we'd end up with the node time from `tables`, thus wrongly increasing the time of the roots of those trees. So, that's not right. Really, we need to add on new nodes above these roots, something like:
```
new_tables = new_ts.tables
for t in new_ts.trees():
    if t.num_roots == 1:
         n = new_tables.nodes.add_row(time=time)
         new_tables.edges.add_row(parent=n, child=t.root, left=t.left, right=t.right)
         new_nodes.append(n)
```

Do you see any issues with this? (It helps to draw the pictures.

Reply:
This looks right to me by inspection, will have to verify with some testing